hash.c

```diff
@@ -2906,8 +2923,37 @@ hash_aset_str(st_data_t *key, st_data_t *val, struct update_arg *arg, int existi
     return hash_aset(key, val, arg, existing);
 }
 
+// Like +hash_aset+, but stashes away the pre-existing value (if any) into +arg->arg+
+static int
+hash_update_value(st_data_t *key, st_data_t *val, struct update_arg *arg, int existing)
+{
+    if (existing) {
+        st_data_t pre_existing_value = *val;
+        *val = arg->arg;
+        arg->arg = pre_existing_value;
+    }
+    else {
+        *val = arg->arg;
+        arg->arg = Qundef;
+    }
+
+    return ST_CONTINUE;
+}
+
+// Like +hash_aset_str+, but stashes away the pre-existing value (if any) into +arg->arg+
+static int
+hash_update_value_str(st_data_t *key, st_data_t *val, struct update_arg *arg, int existing)
+{
+    if (!existing && !RB_OBJ_FROZEN(*key)) {
+        *key = rb_hash_key_str(*key);
+    }
+    return hash_update_value(key, val, arg, existing);
+}
+
```

Comment:
```suggestion
```

Reply:
Why should this be deleted?

(I'm not sure why it should exist, either, I just added it copying from the example of `hash_aset`/`hash_aset_str` 😅 )